notifier/amqp/directdeliverer.go

```diff
@@ -93,8 +93,8 @@ func (d *DirectDeliverer) Deliver(ctx context.Context, _ uuid.UUID) error {
 			be = len(d.n)
 		}
 
-		d.n = d.n[bs:be]
-		err := enc.Encode(&d.n)
+		notificationSlice := d.n[bs:be]
```

Comment:
The idea behind the first, but buggy, way to do this was to avoid allocating a new slice variable over and over...

I think the same can be accomplished, without the bug, like this... 

```
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	var currentBlock []notifier.Notification
	for bs, be := 0, rollup; bs < len(d.n); bs, be = be, be+rollup {
		buf.Reset()
		// if block-end exceeds array bounds, slice block underflow.
		// next block-start will cause loop to exit.
		if be > len(d.n) {
			be = len(d.n)
		}

		currentBlock = d.n[bs:be]
		err := enc.Encode(&currentBlock)
		if err != nil {
			ch.TxRollback()
			return &clairerror.ErrDeliveryFailed{err}
		}

```

in this case, we make one slice var 'currentBlock' and continually reuse it.